src/main/java/org/wso2/carbon/inbound/googlepubsub/GooglePubSubMessageConsumer.java

```diff
@@ -0,0 +1,331 @@
+    private String subscriptionId;
+    private GooglePubSubSubscriptionManager subscriptionManager;
+    private Subscriber subscriber = null;
+    private boolean isConsumed = false;
+    private boolean isPolling = false;
+
+    private long maxOutstandingMessageCount;
+    private long maxOutstandingMessageSize;
+
+    private static final AtomicInteger threadId = new AtomicInteger(1);
+
+    private int concurrentConsumers;
+    private String keyFilePath;
+    private GoogleCredentials credentials;
+    private boolean createSubscriptionOnConnect;
+    private boolean updateSubscriptionIfExists;
+
+    private int executorThreadsPerConsumers;
+    private String endpoint;
+
+    public GooglePubSubMessageConsumer(Properties properties, String name, SynapseEnvironment synapseEnvironment,
+            long scanInterval, String injectingSeq, String onErrorSeq, boolean coordination, boolean sequential) {
+        super(properties, name, synapseEnvironment, scanInterval, injectingSeq, onErrorSeq, coordination, sequential);
+        log.info("Initializing GooglePubSubMessageConsumer");
+        loadConfiguration(properties);
+        if (createSubscriptionOnConnect) {
+            subscriptionManager.createSubscriptionIfNotExists();
+        }
+        if (updateSubscriptionIfExists && !createSubscriptionOnConnect) {
+            subscriptionManager.checkAndUpdateSubscription(ProjectSubscriptionName.of(projectId, subscriptionId));
+        }
+    }
+
+    @Override
+    public Object poll() {
+        log.debug("Polling for messages from Google Pub/Sub...");
+        if (subscriber != null) {
+            isPolling = subscriber.isRunning();
+            if (subscriber.state().equals("FAILED")) {
+                log.info("Subscriber fails.. State: " + subscriber.state());
+                subscriber = null;
+            }
+        }
+        if (!isPolling && subscriber == null) {
+            log.debug("Starting to consume messages from Google Pub/Sub");
+            receiveMessages(projectId, subscriptionId);
+        }
+        return null;
+    }
+
+    @Override
+    public void destroy() {
+        log.info("Stopping the Google Pub/Sub Consumer");
+        if (subscriber != null) {
+            try {
+                subscriber.stopAsync().awaitTerminated(1, TimeUnit.SECONDS);
+            } catch (TimeoutException e) {
+                log.info(
+                        "Timed out while waiting for subscriber to terminate. Subscriber State:  " + subscriber.state());
+            }
+            subscriber = null;
+            log.info("Google Pub/Sub Consumer Stopped ");
+        }
+    }
+
+    private boolean injectMessage(String toStringUtf8, String contentType, MessageContext msgCtx) {
+        AutoCloseInputStream in = new AutoCloseInputStream(new ByteArrayInputStream(toStringUtf8.getBytes()));
+        log.debug("Thread ID: " + Thread.currentThread().getId() + " | Name: " + Thread.currentThread().getName());
```

Comment:
It is best practice to check for whether debug logs are enabled when combining multiple strings. 

        if (log.isDebugEnabled()) {
            // log
        }

This improves the performance because the check happens before the string concatenation.